### PR TITLE
feat: Render PostHog URLs as rich preview chips

### LINKED ISSUE #76253: Render PostHog URLs with rich previews like GitHub URLs
## Context

Requested by Annika in [Slack](https://posthog.slack.com/archives/C09G8Q32R6F/p1777028583118649). GitHub URLs already get rich rendering in PostHog Code (showing repo, issue title, etc.). PostHog URLs are displayed as plain text.

## Problem

PostHog URLs (e.g. links to dashboards, feature flags, insights, recordings) appear as raw URLs with no special formatting or context.

## Expected behavior

PostHog URLs should get the same rich link treatment as GitHub URLs, showing relevant metadata inline (e.g. dashboard name, insight title, feature flag key) so users can understand what a link points to without clicking it.

Image
Image

### PR DESCRIPTION
**Pull request by Basit-Balogun10**
## Problem

When PostHog resource URLs (dashboards, feature flags, insights, experiments, etc.) appear in agent responses or are pasted into the editor, they render as plain text links — unlike GitHub URLs which already get rich chip rendering with icons and resolved titles. This makes it harder to quickly identify and interact with referenced PostHog resources.

Closes PostHog/posthog#76253

## Changes

Mirrors the GitHub URL → chip pipeline for PostHog Cloud URLs across both the MarkdownRenderer (agent messages) and the Tiptap editor (paste handling).

**New files:**
- `posthogUrl.ts` — URL parser supporting both long (`/project/{id}/...`) and short (no project prefix) URL formats. Handles `us.posthog.com`, `eu.posthog.com`, and `localhost:8010`.
- `posthogUrl.test.ts` — 37 tests covering all 13 resource types, both URL formats, edge cases (trailing slashes, query params, fragments), and rejection cases.
- `PostHogRefChip.tsx` — Read-only chip component for agent messages with per-type Phosphor icons.
- `posthogChip.ts` — Chip builder + async title resolution via PostHog API (feature flag names, experiment names, dashboard titles, etc.)

**Modified files:**
- `MarkdownRenderer.tsx` — Adds `SmartPostHogRefChip` that resolves resource titles via `useAuthenticatedQuery`, same pattern as `SmartGithubRefChip`.
- `useTiptapEditor.ts` — Paste-time title resolution with "Loading..." placeholder → resolved label. Works for single URLs, multi-URL pastes, and mixed GitHub + PostHog URL pastes.
- `content.ts` — Persists chip labels in XML (`<feature_flag id="..." label="..." />`) so they survive the user message round-trip. Old messages without a `label` attr gracefully fall back to URL-derived labels.
- `MentionChipNode.ts`, `MentionChipView.tsx` — Extended `ChipType` union and icon map for all new resource types.
- `posthogClient.ts` — 13 new API methods for fetching resource metadata, plus `getDefaultProjectId()` for short URL support.

### Supported resource types

| Resource | URL path segment | Icon | Example |
|----------|-----------------|------|---------|
| Feature Flag | `/feature_flags/{id}` | 🚩 FlagIcon | `Feature Flag #619272 - beta-rollout` |
| Experiment | `/experiments/{id}` | 🧪 FlaskIcon | `Experiment #373424 - signup-test` |
| Insight | `/insights/{id}` | 📈 ChartLineIcon | `Insight KP8iqi6E - Weekly DAU` |
| Dashboard | `/dashboard/{id}` | ◻️ SquaresFourIcon | `Dashboard #944836 - Product KPIs` |
| Error Tracking | `/error_tracking/{id}` | 🐛 BugIcon | `Error abc-def-123 - TypeError...` |
| Recording | `/replay/{id}` | 🎬 VideoIcon | `Recording 019012ab...` |
| Survey | `/surveys/{id}` | 📋 ClipboardTextIcon | `Survey 019d1c79... - NPS Q2` |
| Notebook | `/notebooks/{id}` | 📓 NotebookIcon | `Notebook wkGd - Sprint retro` |
| Cohort | `/cohorts/{id}` | 👥 UsersThreeIcon | `Cohort PostHog/code#55 - Power users` |
| Action | `/data-management/actions/{id}` | ⚡ LightningIcon | `Action PostHog/code#99 - Signed up` |
| Early Access Feature | `/early_access_features/{id}` | 🚀 RocketLaunchIcon | `Early Access Feature abc-123...` |
| Person | `/persons/{id}` | 👤 UserIcon | `Person user_abc123 - jane@example.com` |
| Group | `/groups/{type-index}/{group-key}` | 🏢 BuildingsIcon | `Group 0/my-company-name - Acme Corp` |

All 13 resource types support both long (`/project/{id}/...`) and short (no project prefix) URL formats. Groups use a compound 2-segment ID (`type-index/group-key`).

### What's NOT matched

Org-level and non-resource PostHog URLs render as normal clickable links — no chip, no icon, no metadata resolution:
- `/organization/billing/overview`
- `/settings/...`
- `/feature_flags` (index/list page, no specific resource ID)
- `/project/{id}/feature_flags?search=my-flag` (search/filter page)
- Any URL that doesn't resolve to a specific resource detail page with an ID

These are intentionally excluded — chips are only shown when we can identify a discrete resource to link to.

## How did you test this?

**Automated:**
- 37 new unit tests for the URL parser (all pass)
- 25 existing content.ts round-trip tests (all pass)
- Full typecheck passes
- Biome lint passes

**Manual testing steps**

**Demo 1 — Resource chip rendering & title resolution**

Pasted/referenced all 13 supported resource types (Feature Flag → Group) covering both long (`/project/{id}/...`) and short (no project prefix) URL formats.

- URLs with real, valid resource IDs belonging to the authenticated org resolve and display the resource title in the chip (e.g. `Feature Flag #619272 - beta-rollout`)
- URLs with fake/test IDs, or IDs belonging to orgs the user doesn't have access to, render the chip with the ID only — no title — which is the expected graceful fallback behavior

[valid-posthog-resource-urls-rendering-posthog-code.webm](https://github.com/user-attachments/assets/155ef9d1-b07a-4eef-a097-63102f264377)

**Demo 2 — Non-resource URLs & GitHub interop**

Verified that org-level and list/index PostHog URLs are intentionally excluded from chip rendering and fall back to plain clickable links:

- `https://us.posthog.com/feature_flags` → plain link
- `https://us.posthog.com/organization/billing/overview` → plain link
- `https://us.posthog.com/data-warehouse` → plain link
- `https://us.posthog.com/project/99271/feature_flags?search=my-flag` → plain link

Also confirmed that existing GitHub URL chip rendering is unaffected:

- `https://github.com/PostHog/code/issues/2360` → renders as GitHub chip as before

[org-level-and-non-resource-urls-shouldnt-render-as-chips-posthog-code.webm](https://github.com/user-attachments/assets/6e2288a5-87fb-4a0b-bdc5-1eef948ff8cc)

## Publish to changelog?

no

### CODE REVIEW

File: packages/api-client/src/posthog-client.ts [hunk 1]
```diff
@@ -218,7 +218,7 @@ export interface SignalSourceConfig {
   status: "running" | "completed" | "failed" | null;
 }
 
-// ── Signals scouts ───────────────────────────────────────────────────────────
+// â”€â”€ Signals scouts â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€
```
Comment (charlesvien) [inline packages/api-client/src/posthog-client.ts]: ??

**Review by charlesvien — COMMENTED** [review]
Can you fix the  â€”'s here

File: packages/api-client/src/posthog-client.ts (hunk 1 above)
Comment (Basit-Balogun10) [reply packages/api-client/src/posthog-client.ts]: Oh looks like some (windows) encoding corruption (possibly from the last conflcit resolution I did). Will address in a bit. Thank you

Comment (Basit-Balogun10) [reply packages/api-client/src/posthog-client.ts]: It's done, re-requesting review (thanks!)

**Comment by Basit-Balogun10** [comment]
Can confirm there are currently no regressions  upon conflict resolution :

RCsQUvEY06Gf

image

**Comment by Basit-Balogun10** [comment]
Now ported to the new ChatThread

UzBelJPMtGVA

kmuBeJE75Y8A

File: packages/core/src/message-editor/content.ts [hunk 2]
```diff
@@ -59,11 +68,17 @@ export function contentToXml(content: EditorContent): string {
...
-        return `<feature_flag id="${escapedId}" />`;
+      case "dashboard":
+      case "recording":
+      case "error_tracking":
+      case "survey":
+      case "notebook":
+      case "cohort":
+      case "action":
+      case "early_access_feature":
+        return `<${chip.type} id="${escapedId}" label="${escapeXmlAttr(chip.label)}" />`;
```
[the remark this answers is not in the document]
Comment (Basit-Balogun10) [reply packages/core/src/message-editor/content.ts]: Fixed — `contentToXml` now runs the label through `stripPlaceholderLabelSuffix` before writing the `label` attribute, so a message submitted mid-resolve persists the clean base label (e.g. `Feature Flag #42`) instead of `... - Loading...`. Added round-trip tests in `content.test.ts` covering both the placeholder-strip and the fully-resolved cases.

File: apps/code/src/renderer/api/posthogClient.ts [hunk 3]
```diff
@@ -2887,4 +2891,180 @@ export class PostHogAPIClient {
...
+    const response = await this.api.fetcher.fetch({
+      method: "get",
+      url,
+      path: urlPath,
+    });
+    if (!response.ok) return null;
+    const data = (await response.json()) as { name?: string };
+    return { name: data.name ?? "" };
+  }
 }
```
[the remark this answers is not in the document]
Comment (Basit-Balogun10) [reply apps/code/src/renderer/api/posthogClient.ts:6369]: Addressed — extracted the shared fetch/null/JSON boilerplate into a private `fetchProjectResourceById(resourcePath, projectId, resourceId)` helper. The 11 by-id getters now reduce to a one-line field mapping. `getPerson`/`getGroup` stay separate since they use query-param URLs and array-shaped results.

File: packages/ui/src/features/message-editor/tiptap/MentionChipView.tsx [hunk 4]
```diff
@@ -82,17 +98,24 @@ function DefaultChip({
   const isFile = type === "file";
   const isFolder = type === "folder";
   const isGithubRef = type === "github_issue" || type === "github_pr";
-  const canOpenUrl = isGithubRef && /^https:\/\//.test(id);
+  const isPostHogRef =
+    type !== "file" &&
+    type !== "folder" &&
+    type !== "command" &&
+    type !== "error" &&
+    !isGithubRef;
```
[the remark this answers is not in the document]
Comment (Basit-Balogun10) [reply packages/ui/src/features/message-editor/tiptap/MentionChipView.tsx]: Addressed — replaced the exclusion-based check with a positive `isPostHogResourceType(type)` membership test. It's backed by `POSTHOG_RESOURCE_TYPES` in `posthogUrl.ts`, which is now the single source of truth (the `PostHogResourceType` union is derived from it), so a future chip type is never silently treated as a clickable PostHog chip.

File: packages/api-client/src/posthog-client.ts [hunk 5]
```diff
@@ -600,6 +600,10 @@ export class PostHogAPIClient {
     throw new Error("No team found for user");
   }
 
+  async getDefaultProjectId(): Promise<number> {
+    return this.getTeamId();
+  }
```
[the remark this answers is not in the document]
Comment (Basit-Balogun10) [reply packages/api-client/src/posthog-client.ts]: Dropped the redundant `async` (it now just forwards the already-Promise result). Kept the method itself: `getTeamId()` is private, so this public alias is how `posthogChip.ts` reaches it under a clearer name.

**Comment by charlesvien** [comment]
Closing stale PRs ahead of Friday's monorepo migration. If this is a mistake, rebase and reopen for review.
